File: geomath/polygon_utils.py
"""Polygon helpers for the geomath demo project."""

from __future__ import annotations

from typing import Sequence

from .vector2d import Vector2D


def area(vertices: Sequence[Vector2D] | None) -> float:
    """Return the non-negative area of a simple polygon.

    The vertices are processed using the shoelace formula.  Returns ``0.0``
    when fewer than three vertices are provided.
    """

    if not vertices or len(vertices) < 3:
        return 0.0

    s = 0.0
    n = len(vertices)
    for i in range(n):
        a = vertices[i]
        b = vertices[(i + 1) % n]
        s += a.x * b.y - a.y * b.x
    return abs(s) * 0.5


def is_convex(vertices: Sequence[Vector2D] | None) -> bool:
    """Return ``True`` when the polygon defined by ``vertices`` is convex."""

    if not vertices or len(vertices) < 3:
        return False

    n = len(vertices)
    sign = 0
    for i in range(n):
        a = vertices[i]
        b = vertices[(i + 1) % n]
        c = vertices[(i + 2) % n]
        ab = b.subtract(a)
        bc = c.subtract(b)
        cross = ab.cross(bc)
        if cross != 0.0:
            curr = 1 if cross > 0 else -1
            if sign == 0:
                sign = curr
            elif sign != curr:
                return False
    return True

File: src/main/java/com/example/geomath/PolygonUtils.java
package com.example.geomath;

import java.util.List;

public final class PolygonUtils {
    private PolygonUtils() {}

    /**
     * Non-negative polygon area via the shoelace formula.
     * Assumes points form a simple (non self-intersecting) polygon.
     */
    public static double area(List<Vector2D> pts) {
        if (pts == null || pts.size() < 3) return 0.0;
        double s = 0.0;
        int n = pts.size();
        for (int i = 0; i < n; i++) {
            Vector2D a = pts.get(i);
            Vector2D b = pts.get((i + 1) % n);
            s += a.x() * b.y() - a.y() * b.x();
        }
        return Math.abs(s) * 0.5;
    }

    /**
     * Basic convexity check: polygon edges should not change the sign
     * of successive cross products. Colinear steps are allowed.
     * Returns false for fewer than 3 points.
     */
    public static boolean isConvex(List<Vector2D> pts) {
        if (pts == null || pts.size() < 3) return false;

        int n = pts.size();
        int sign = 0; // -1, 0, +1
        for (int i = 0; i < n; i++) {
            Vector2D a = pts.get(i);
            Vector2D b = pts.get((i + 1) % n);
            Vector2D c = pts.get((i + 2) % n);
            Vector2D ab = b.subtract(a);
            Vector2D bc = c.subtract(b);
            double cross = ab.cross(bc);
            if (cross != 0.0) {
                int curr = cross > 0 ? 1 : -1;
                if (sign == 0) {
                    sign = curr;
                } else if (sign != curr) {
                    return false;
                }
            }
        }
        return true;
    }
}
